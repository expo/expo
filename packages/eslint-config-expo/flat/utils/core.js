const expo = require('eslint-plugin-expo');
const importPlugin = require('eslint-plugin-import');
const { jsExtensions } = require('./extensions');

const globals = require('globals');

module.exports = [
  importPlugin.flatConfigs.recommended,
  importPlugin.flatConfigs.errors,
  {
    plugins: {
      expo,
    },

    languageOptions: {
      globals: {
        console: 'readonly',
        exports: false,
        global: false,
        module: false,
        require: false,
      },

      ecmaVersion: 2022,
      sourceType: 'module',

      parserOptions: {
        ecmaFeatures: {
          impliedStrict: true,
          jsx: true,
        },
      },
    },

    settings: {
      'import/ignore': ['node_modules[\\\\/]+@?react-native'],

      'import/extensions': jsExtensions,
      'import/resolver': {
        node: { extensions: jsExtensions },
        typescript: true,
      },
    },

    rules: {
      eqeqeq: ['warn', 'smart'],
      'no-dupe-args': 'error',
      'no-dupe-class-members': 'error',
      'no-dupe-keys': 'error',
      'no-duplicate-case': 'error',
      'no-empty-character-class': 'warn',
      'no-empty-pattern': 'warn',
      'no-extend-native': 'warn',
      'no-extra-bind': 'warn',
      'no-redeclare': 'warn',
      'no-undef': 'error',
      'no-unreachable': 'warn',
      'no-unsafe-negation': 'warn',

      'no-unused-expressions': [
        'warn',
        {
          allowShortCircuit: true,
          enforceForJSX: true,
        },
      ],

      'no-unused-labels': 'warn',

      'no-unused-vars': [
        'warn',
        {
          vars: 'all',
          args: 'none',
          ignoreRestSiblings: true,
          caughtErrors: 'all',
          caughtErrorsIgnorePattern: '^_',
        },
      ],

      'no-with': 'warn',
      'unicode-bom': ['warn', 'never'],
      'use-isnan': 'error',
      'valid-typeof': 'error',
      'import/first': 'warn',
      'import/default': 'off',
    },
  },
  {
    files: ['**/*.d.ts'],

    rules: {
      'import/order': 'off',
    },
  },
  {
    files: ['**/metro.config.js'],

    languageOptions: {
      globals: globals.node,
    },
  },
];
