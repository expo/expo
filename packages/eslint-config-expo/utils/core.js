const { jsExtensions } = require('./extensions');

module.exports = {
  parserOptions: {
    sourceType: 'module',
    ecmaVersion: 2022,
    ecmaFeatures: { impliedStrict: true, jsx: true },
  },
  env: { es2022: true },
  globals: {
    console: 'readonly',
    exports: false,
    global: false,
    module: false,
    require: false,
  },
  extends: ['plugin:import/errors'],
  plugins: ['import', 'expo'],
  rules: {
    eqeqeq: ['warn', 'smart'],
    'no-dupe-args': 'error',
    'no-dupe-class-members': 'error',
    'no-dupe-keys': 'error',
    'no-duplicate-case': 'error',
    'no-empty-character-class': 'warn',
    'no-empty-pattern': 'warn',
    'no-extend-native': 'warn',
    'no-extra-bind': 'warn',
    'no-redeclare': 'warn',
    'no-undef': 'error',
    'no-unreachable': 'warn',
    'no-unsafe-negation': 'warn',
    'no-unused-expressions': ['warn', { allowShortCircuit: true, enforceForJSX: true }],
    'no-unused-labels': 'warn',
    'no-unused-vars': [
      'warn',
      {
        vars: 'all',
        args: 'none',
        ignoreRestSiblings: true,
        caughtErrors: 'all',
        caughtErrorsIgnorePattern: '^_',
      },
    ],
    'no-with': 'warn',
    'unicode-bom': ['warn', 'never'],
    'use-isnan': 'error',
    'valid-typeof': 'error',
    'import/first': 'warn', //keep
    'import/default': 'off',
  },
  settings: {
    'import/extensions': jsExtensions,
    'import/ignore': [
      // react-native's main module is Flow, not JavaScript, and raises parse errors. Additionally,
      // several other react-native-related packages still publish Flow code as their main source.
      'node_modules[\\\\/]+@?react-native',
    ],
    'import/resolver': {
      node: { extensions: jsExtensions },
      typescript: true
    },
  },
  overrides: [
    {
      files: ['*.d.ts'],
      rules: {
        'import/order': 'off',
      },
    },
  ],
};
