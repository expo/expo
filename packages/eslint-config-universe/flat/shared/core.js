const { defineConfig } = require('eslint/config');
const importPlugin = require('eslint-plugin-import');
const nodePlugin = require('eslint-plugin-n');
const globals = require('globals');

const { jsExtensions } = require('./extensions');

module.exports = defineConfig([
  importPlugin.flatConfigs.recommended,
  {
    plugins: {
      n: nodePlugin,
    },

    languageOptions: {
      globals: {
        ...globals.jest,
        console: false,
        exports: false,
        global: false,
        module: false,
        require: false,
      },

      ecmaVersion: 2022,
      sourceType: 'module',

      parserOptions: {
        ecmaFeatures: {
          impliedStrict: true,
          jsx: true,
        },
      },
    },

    settings: {
      'import/extensions': jsExtensions,
      'import/ignore': ['node_modules[\\\\/]+@?react-native'],

      'import/resolver': {
        node: {
          extensions: jsExtensions,
        },
      },
    },

    rules: {
      'array-bracket-spacing': ['warn', 'never'],

      'arrow-spacing': [
        'warn',
        {
          before: true,
          after: true,
        },
      ],

      curly: ['warn', 'all'],
      'block-spacing': ['warn', 'always'],
      'comma-dangle': ['warn', 'always-multiline'],

      'comma-spacing': [
        'warn',
        {
          before: false,
          after: true,
        },
      ],

      'comma-style': ['warn', 'last'],
      'computed-property-spacing': ['warn', 'never'],
      'constructor-super': 'warn',
      'dot-location': ['warn', 'property'],
      'eol-last': 'warn',
      eqeqeq: ['warn', 'smart'],
      'func-call-spacing': ['warn', 'never'],
      'generator-star-spacing': ['warn', 'after'],
      'getter-return': 'warn',
      'jsx-quotes': ['warn', 'prefer-double'],
      'new-parens': 'warn',
      'no-alert': 'off',
      'no-array-constructor': 'warn',
      'no-caller': 'warn',
      'no-case-declarations': 'warn',
      'no-compare-neg-zero': 'warn',
      'no-cond-assign': 'warn',
      'no-const-assign': 'error',

      'no-constant-condition': [
        'warn',
        {
          checkLoops: false,
        },
      ],

      'no-control-regex': 'off',
      'no-debugger': 'warn',
      'no-delete-var': 'error',
      'no-dupe-args': 'error',
      'no-dupe-class-members': 'error',
      'no-dupe-keys': 'error',
      'no-duplicate-case': 'error',
      'no-empty-character-class': 'warn',
      'no-empty-pattern': 'warn',
      'no-eval': 'warn',
      'no-ex-assign': 'warn',
      'no-extend-native': 'warn',
      'no-extra-bind': 'warn',
      'no-extra-boolean-cast': 'warn',
      'no-extra-semi': 'warn',
      'no-fallthrough': 'warn',
      'no-floating-decimal': 'warn',
      'no-func-assign': 'error',
      'no-global-assign': 'warn',
      'no-implied-eval': 'warn',
      'no-inner-declarations': 'warn',
      'no-invalid-regexp': 'error',
      'no-irregular-whitespace': 'warn',
      'no-iterator': 'warn',
      'no-label-var': 'warn',

      'no-labels': [
        'warn',
        {
          allowLoop: true,
          allowSwitch: true,
        },
      ],

      'no-lone-blocks': 'warn',
      'no-multi-assign': 'warn',
      'no-new': 'warn',
      'no-new-func': 'warn',
      'no-object-constructor': 'warn',
      'no-new-native-nonconstructor': 'error',
      'no-obj-calls': 'warn',
      'no-octal': 'warn',
      'no-octal-escape': 'warn',
      'no-mixed-spaces-and-tabs': 'warn',
      'no-proto': 'warn',
      'no-redeclare': 'warn',
      'no-return-assign': 'warn',
      'no-script-url': 'warn',
      'no-self-assign': 'warn',
      'no-self-compare': 'warn',
      'no-sequences': 'warn',
      'no-shadow-restricted-names': 'warn',
      'no-sparse-arrays': 'warn',
      'no-this-before-super': 'warn',
      'no-throw-literal': 'warn',
      'no-trailing-spaces': 'warn',
      'no-undef': 'error',
      'no-unexpected-multiline': 'warn',
      'no-unneeded-ternary': 'warn',
      'no-unreachable': 'warn',
      'no-unsafe-negation': 'warn',

      'no-unused-expressions': [
        'warn',
        {
          allowShortCircuit: true,
          enforceForJSX: true,
        },
      ],

      'no-unused-labels': 'warn',

      'no-unused-vars': [
        'warn',
        {
          vars: 'all',
          args: 'none',
          ignoreRestSiblings: true,
          caughtErrors: 'all',
          caughtErrorsIgnorePattern: '^_',
        },
      ],

      'no-useless-computed-key': 'warn',
      'no-useless-concat': 'warn',
      'no-useless-constructor': 'warn',
      'no-useless-escape': 'warn',
      'no-useless-rename': 'warn',
      'no-useless-return': 'warn',
      'no-var': 'warn',
      'no-void': 'warn',
      'no-whitespace-before-property': 'warn',
      'no-with': 'warn',
      'object-shorthand': 'warn',

      'operator-linebreak': [
        'warn',
        'after',
        {
          overrides: {
            '?': 'before',
            ':': 'before',
          },
        },
      ],

      'prefer-const': [
        'warn',
        {
          destructuring: 'all',
          ignoreReadBeforeAssign: true,
        },
      ],

      'prefer-promise-reject-errors': 'warn',
      'prefer-rest-params': 'warn',
      'prefer-spread': 'warn',
      radix: 'warn',
      'rest-spread-spacing': ['warn', 'never'],
      semi: 'warn',

      'semi-spacing': [
        'warn',
        {
          before: false,
          after: true,
        },
      ],

      'semi-style': ['warn', 'last'],
      'space-before-blocks': ['warn', 'always'],

      'space-before-function-paren': [
        'warn',
        {
          anonymous: 'never',
          named: 'never',
        },
      ],

      'space-in-parens': ['warn', 'never'],
      'space-infix-ops': 'warn',

      'switch-colon-spacing': [
        'warn',
        {
          before: false,
          after: true,
        },
      ],

      'template-curly-spacing': ['warn', 'never'],
      'template-tag-spacing': ['warn', 'never'],
      'unicode-bom': ['warn', 'never'],
      'use-isnan': 'error',
      'valid-typeof': 'error',
      'yield-star-spacing': ['warn', 'after'],

      yoda: [
        'warn',
        'never',
        {
          exceptRange: true,
        },
      ],

      'import/default': 'off',
      'import/export': 'error',
      'import/first': 'warn',

      'import/namespace': [
        'error',
        {
          allowComputed: true,
        },
      ],

      'import/no-duplicates': 'error',

      // TODO (Kadi): Enable this. Disabling for now because import/recommended adds it, but we didn't use to have it enabled
      'import/no-unresolved': 'off',

      'import/order': [
        'warn',
        {
          groups: [['builtin', 'external'], 'internal', ['parent', 'index', 'sibling']],
          'newlines-between': 'always',

          alphabetize: {
            order: 'asc',
          },
        },
      ],

      'n/handle-callback-err': ['warn', '^(e|err|error|.+Error)$'],
      'n/no-new-require': 'warn',
    },
  },
  {
    files: ['**/*.d.ts'],

    rules: {
      'import/order': 'off',
    },
  },
]);
