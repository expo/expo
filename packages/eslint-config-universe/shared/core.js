const { jsExtensions } = require('./extensions');

module.exports = {
  parserOptions: {
    sourceType: 'module',
    ecmaVersion: 2022,
    ecmaFeatures: { impliedStrict: true, jsx: true },
  },
  env: { es2022: true, jest: true },
  globals: {
    console: false,
    exports: false,
    global: false,
    module: false,
    require: false,
  },
  plugins: ['import', 'node'],
  rules: {
    'array-bracket-spacing': ['warn', 'never'],
    'arrow-spacing': ['warn', { before: true, after: true }],
    curly: ['warn', 'all'],
    'block-spacing': ['warn', 'always'],
    'comma-dangle': ['warn', 'always-multiline'],
    'comma-spacing': ['warn', { before: false, after: true }],
    'comma-style': ['warn', 'last'],
    'computed-property-spacing': ['warn', 'never'],
    'constructor-super': 'warn',
    'dot-location': ['warn', 'property'],
    'eol-last': 'warn',
    eqeqeq: ['warn', 'smart'],
    'func-call-spacing': ['warn', 'never'],
    'generator-star-spacing': ['warn', 'after'],
    'getter-return': 'warn',
    'jsx-quotes': ['warn', 'prefer-double'],
    'new-parens': 'warn',
    'no-alert': 'off',
    'no-array-constructor': 'warn',
    'no-caller': 'warn',
    'no-case-declarations': 'warn',
    'no-compare-neg-zero': 'warn',
    'no-cond-assign': 'warn',
    'no-const-assign': 'error',
    'no-constant-condition': ['warn', { checkLoops: false }],
    'no-control-regex': 'off',
    'no-debugger': 'warn',
    'no-delete-var': 'error',
    'no-dupe-args': 'error',
    'no-dupe-class-members': 'error',
    'no-dupe-keys': 'error',
    'no-duplicate-case': 'error',
    'no-empty-character-class': 'warn',
    'no-empty-pattern': 'warn',
    'no-eval': 'warn',
    'no-ex-assign': 'warn',
    'no-extend-native': 'warn',
    'no-extra-bind': 'warn',
    'no-extra-boolean-cast': 'warn',
    'no-extra-semi': 'warn',
    'no-fallthrough': 'warn',
    'no-floating-decimal': 'warn',
    'no-func-assign': 'error',
    'no-global-assign': 'warn',
    'no-implied-eval': 'warn',
    'no-inner-declarations': 'warn',
    'no-invalid-regexp': 'error',
    'no-irregular-whitespace': 'warn',
    'no-iterator': 'warn',
    'no-label-var': 'warn',
    'no-labels': ['warn', { allowLoop: true, allowSwitch: true }],
    'no-lone-blocks': 'warn',
    'no-multi-assign': 'warn',
    'no-new': 'warn',
    'no-new-func': 'warn',
    'no-new-object': 'warn',
    'no-new-symbol': 'error',
    'no-obj-calls': 'warn',
    'no-octal': 'warn',
    'no-octal-escape': 'warn',
    'no-mixed-spaces-and-tabs': 'warn',
    'no-proto': 'warn',
    'no-redeclare': 'warn',
    'no-return-assign': 'warn',
    'no-script-url': 'warn',
    'no-self-assign': 'warn',
    'no-self-compare': 'warn',
    'no-sequences': 'warn',
    'no-shadow-restricted-names': 'warn',
    'no-sparse-arrays': 'warn',
    'no-this-before-super': 'warn',
    'no-throw-literal': 'warn',
    'no-trailing-spaces': 'warn',
    'no-undef': 'error',
    'no-unexpected-multiline': 'warn',
    'no-unneeded-ternary': 'warn',
    'no-unreachable': 'warn',
    'no-unsafe-negation': 'warn',
    'no-unused-expressions': ['warn', { allowShortCircuit: true, enforceForJSX: true }],
    'no-unused-labels': 'warn',
    'no-unused-vars': [
      'warn',
      {
        vars: 'all',
        args: 'none',
        ignoreRestSiblings: true,
        caughtErrors: 'all',
        caughtErrorsIgnorePattern: '^_',
      },
    ],
    'no-useless-computed-key': 'warn',
    'no-useless-concat': 'warn',
    'no-useless-constructor': 'warn',
    'no-useless-escape': 'warn',
    'no-useless-rename': 'warn',
    'no-useless-return': 'warn',
    'no-var': 'warn',
    'no-void': 'warn',
    'no-whitespace-before-property': 'warn',
    'no-with': 'warn',
    'object-shorthand': 'warn',
    'operator-linebreak': ['warn', 'after', { overrides: { '?': 'before', ':': 'before' } }],
    'prefer-const': ['warn', { destructuring: 'all', ignoreReadBeforeAssign: true }],
    'prefer-promise-reject-errors': 'warn',
    'prefer-rest-params': 'warn',
    'prefer-spread': 'warn',
    radix: 'warn',
    'rest-spread-spacing': ['warn', 'never'],
    semi: 'warn',
    'semi-spacing': ['warn', { before: false, after: true }],
    'semi-style': ['warn', 'last'],
    'space-before-blocks': ['warn', 'always'],
    'space-before-function-paren': ['warn', { anonymous: 'never', named: 'never' }],
    'space-in-parens': ['warn', 'never'],
    'space-infix-ops': 'warn',
    'switch-colon-spacing': ['warn', { before: false, after: true }],
    'template-curly-spacing': ['warn', 'never'],
    'template-tag-spacing': ['warn', 'never'],
    'unicode-bom': ['warn', 'never'],
    'use-isnan': 'error',
    'valid-typeof': 'error',
    'yield-star-spacing': ['warn', 'after'],
    yoda: ['warn', 'never', { exceptRange: true }],

    'import/default': 'off',
    'import/export': 'error',
    'import/first': 'warn',
    'import/namespace': ['error', { allowComputed: true }],
    'import/no-duplicates': 'error',
    'import/order': [
      'warn',
      {
        groups: [['builtin', 'external'], 'internal', ['parent', 'index', 'sibling']],
        'newlines-between': 'always',
        alphabetize: {
          order: 'asc',
        },
      },
    ],

    'node/handle-callback-err': ['warn', '^(e|err|error|.+Error)$'],
    'node/no-new-require': 'warn',
  },
  settings: {
    'import/extensions': jsExtensions,
    'import/ignore': [
      // react-native's main module is Flow, not JavaScript, and raises parse errors. Additionally,
      // several other react-native-related packages still publish Flow code as their main source.
      'node_modules[\\\\/]+@?react-native',
    ],
    'import/resolver': {
      node: { extensions: jsExtensions },
    },
  },
  overrides: [
    {
      files: ['*.d.ts'],
      rules: {
        'import/order': 'off',
      },
    },
  ],
};
