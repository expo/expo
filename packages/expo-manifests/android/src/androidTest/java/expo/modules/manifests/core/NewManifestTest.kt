package expo.modules.manifests.core

import androidx.test.internal.runner.junit4.AndroidJUnit4ClassRunner
import org.json.JSONObject
import org.junit.Assert
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4ClassRunner::class)
class NewManifestTest {
  @Test
  @Throws(Exception::class)
  fun testGetSDKVersionNullable_ValidCases() {
    val runtimeVersion = "exposdk:39.0.0"
    val manifestJson =
      "{\"runtimeVersion\":\"$runtimeVersion\"}"
    val manifest = NewManifest(JSONObject(manifestJson))
    Assert.assertEquals(manifest.getExpoGoSDKVersion(), "39.0.0")
  }

  @Test
  @Throws(Exception::class)
  fun testGetSDKVersionNullable_ValidCaseUnversioned() {
    val runtimeVersion = "exposdk:UNVERSIONED"
    val manifestJson =
      "{\"runtimeVersion\":\"$runtimeVersion\"}"
    val manifest = NewManifest(JSONObject(manifestJson))
    Assert.assertEquals(manifest.getExpoGoSDKVersion(), "UNVERSIONED")
  }

  @Test
  @Throws(Exception::class)
  fun testGetSDKVersionNullable_NotSDKRuntimeVersionCases() {
    val runtimeVersions = listOf(
      "exposdk:123",
      "exposdkd:39.0.0",
      "exposdk:hello",
      "bexposdk:39.0.0",
      "exposdk:39.0.0-beta.0",
      "exposdk:39.0.0-alpha.256"
    )
    runtimeVersions.forEach { runtimeVersion ->
      val manifestJson =
        "{\"runtimeVersion\":\"$runtimeVersion\"}"
      val manifest = NewManifest(JSONObject(manifestJson))
      Assert.assertNull(manifest.getExpoGoSDKVersion())
    }
  }
}
