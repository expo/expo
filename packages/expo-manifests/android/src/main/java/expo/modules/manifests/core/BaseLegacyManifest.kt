package expo.modules.manifests.core

import expo.modules.jsonutils.getNullable
import expo.modules.jsonutils.require
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

abstract class BaseLegacyManifest(json: JSONObject) : Manifest(json) {
  override fun getStableLegacyID(): String = json.getNullable("originalFullName") ?: getLegacyID()

  override fun getScopeKey(): String = json.getNullable("scopeKey") ?: getStableLegacyID()

  override fun getEASProjectID(): String? = json.getNullable("projectId")

  override fun getAssets(): JSONArray? = json.getNullable("assets")

  @Throws(JSONException::class)
  override fun getBundleURL(): String = json.require("bundleUrl")

  override fun getExpoGoSDKVersion(): String? = json.getNullable("sdkVersion")

  override fun getExpoGoConfigRootObject(): JSONObject? {
    return json
  }

  override fun getExpoClientConfigRootObject(): JSONObject? {
    return json
  }

  override fun getSlug(): String? = json.getNullable("slug")

  override fun getAppKey(): String? = json.getNullable("appKey")

  fun getCommitTime(): String? = json.getNullable("commitTime")
}
