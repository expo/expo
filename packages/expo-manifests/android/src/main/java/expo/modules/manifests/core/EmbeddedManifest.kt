package expo.modules.manifests.core

import expo.modules.jsonutils.getNullable
import expo.modules.jsonutils.require
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject

class EmbeddedManifest(json: JSONObject) : Manifest(json) {
  /**
   * A UUID for this manifest.
   */
  @Throws(JSONException::class)
  fun getID(): String = json.require("id")

  @Throws(JSONException::class)
  fun getCommitTimeLong(): Long = json.require("commitTime")

  override fun getStableLegacyID(): String = json.getNullable("originalFullName") ?: getLegacyID()

  override fun getScopeKey(): String = json.getNullable("scopeKey") ?: getStableLegacyID()

  override fun getEASProjectID(): String? = json.getNullable("projectId")

  override fun getAssets(): JSONArray? = json.getNullable("assets")

  @Throws(JSONException::class)
  override fun getBundleURL(): String = json.require("bundleUrl")

  override fun getExpoGoSDKVersion(): String? = json.getNullable("sdkVersion")

  override fun getExpoGoConfigRootObject(): JSONObject {
    return json
  }

  override fun getExpoClientConfigRootObject(): JSONObject {
    return json
  }

  override fun getSlug(): String? = json.getNullable("slug")

  override fun getAppKey(): String? = json.getNullable("appKey")
}
