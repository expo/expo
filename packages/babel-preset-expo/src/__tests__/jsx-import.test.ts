import * as babel from '@babel/core';

import preset from '..';

function getCaller(props: Record<string, string | boolean>): babel.TransformCaller {
  return props as unknown as babel.TransformCaller;
}

const DEF_OPTIONS = {
  // Ensure this is absolute to prevent the filename from being converted to absolute and breaking CI tests.
  filename: '/unknown',
  babelrc: false,
  presets: [preset],
  sourceMaps: true,
  configFile: false,
  compact: false,
  comments: true,
  retainLines: true,
  caller: getCaller({ name: 'metro', engine: 'hermes', platform: 'ios' }),
};

it(`compiles React auto jsx import`, () => {
  const options = {
    ...DEF_OPTIONS,
    caller: getCaller({ name: 'metro', engine: 'hermes', platform: 'ios', isDev: true }),
  };

  const sourceCode = `
function App() {
    return <div />
}`;
  expect(babel.transform(sourceCode, options)!.code).toMatchInlineSnapshot(`
    "var _jsxDevRuntime = require("react/jsx-dev-runtime");var _jsxFileName = "/unknown";
    function App() {
      return /*#__PURE__*/(0, _jsxDevRuntime.jsxDEV)("div", {}, void 0, false, { fileName: _jsxFileName, lineNumber: 3, columnNumber: 12 }, this);
    }"
  `);

  const productionOptions = {
    ...DEF_OPTIONS,
    caller: getCaller({ name: 'metro', engine: 'hermes', platform: 'ios', isDev: false }),
  };
  expect(babel.transform(sourceCode, productionOptions)!.code).toMatchInlineSnapshot(`
    "var _jsxRuntime = require("react/jsx-runtime");
    function App() {
      return /*#__PURE__*/(0, _jsxRuntime.jsx)("div", {});
    }"
  `);
});

it(`transforms React display name`, () => {
  const options = {
    ...DEF_OPTIONS,
    caller: getCaller({ name: 'metro', engine: 'hermes', platform: 'ios', isDev: true }),
  };

  // Ensure no duplication
  const sourceCode = `
  var bar = createReactClass({});
  `;
  expect(babel.transform(sourceCode, options)!.code).toMatchInlineSnapshot(`
    "
    var bar = createReactClass({ displayName: "bar" });"
  `);
});

describe('classic runtime', () => {
  // No React import...
  const sourceCode = `
  import { View } from 'react-native';
  export default function App() {
    return <View />
  }`;

  it(`compiles for Webpack in dev`, () => {
    const options = {
      ...DEF_OPTIONS,
      presets: [[preset, { jsxRuntime: 'classic' }]],
      caller: getCaller({
        name: 'babel-loader',
        isDev: true,
      }),
    };

    const code = babel.transform(sourceCode, options)!.code;

    expect(code).not.toMatch(/"react\/jsx-runtime"/);
    expect(code).not.toMatch(/_jsx\(View/);
    expect(code).toMatchInlineSnapshot(`
      "var _jsxFileName = "/unknown";import View from "react-native-web/dist/exports/View";

      export default function App() {
        return /*#__PURE__*/React.createElement(View, { __self: this, __source: { fileName: _jsxFileName, lineNumber: 4, columnNumber: 12 } });
      }"
    `);
  });
  it(`compiles for Webpack in prod`, () => {
    const options = {
      ...DEF_OPTIONS,
      presets: [[preset, { jsxRuntime: 'classic' }]],
      caller: getCaller({
        name: 'babel-loader',
        isDev: false,
      }),
    };

    const code = babel.transform(sourceCode, options)!.code;

    expect(code).not.toMatch(/"react\/jsx-runtime"/);
    expect(code).not.toMatch(/_jsx\(View/);
    expect(code).toMatchInlineSnapshot(`
      "import View from "react-native-web/dist/exports/View";

      export default function App() {
        return /*#__PURE__*/React.createElement(View, null);
      }"
    `);
  });

  ['ios', 'web'].forEach((platform) => {
    it(`compiles for Metro ${platform} in dev`, () => {
      const options = {
        ...DEF_OPTIONS,
        presets: [[preset, { jsxRuntime: 'classic' }]],
        caller: getCaller({
          name: 'metro',
          platform,
          isDev: true,
        }),
      };

      const code = babel.transform(sourceCode, options)!.code;

      expect(code).not.toMatch(/"react\/jsx-runtime"/);
      // Format is a little different for Metro
      expect(code).not.toMatch(/_jsxRuntime\.jsx/);
      expect(code).toMatchSnapshot();
    });
    it(`compiles for Metro ${platform} in prod`, () => {
      const options = {
        ...DEF_OPTIONS,
        presets: [[preset, { jsxRuntime: 'classic' }]],
        caller: getCaller({
          name: 'metro',
          platform,
          isDev: false,
        }),
      };

      const code = babel.transform(sourceCode, options)!.code;

      expect(code).not.toMatch(/"react\/jsx-runtime"/);
      // Format is a little different for Metro
      expect(code).not.toMatch(/_jsxRuntime\.jsx/);
      expect(code).toMatchSnapshot();
    });
  });
});

describe('auto runtime (default)', () => {
  // No React import...
  const sourceCode = `
  import { View } from 'react-native';
  export default function App() {
    return <View />
  }`;

  it(`compiles for Webpack in dev`, () => {
    const options = {
      ...DEF_OPTIONS,
      presets: [[preset, { jsxRuntime: 'automatic' }]],
      caller: getCaller({
        name: 'babel-loader',
        isDev: true,
      }),
    };

    const code = babel.transform(sourceCode, options)!.code;

    expect(code).toMatch(/"react\/jsx-dev-runtime"/);
    expect(code).not.toMatch(/_jsx\(View/);
    expect(code).toMatchInlineSnapshot(`
      "var _jsxFileName = "/unknown";import View from "react-native-web/dist/exports/View";import { jsxDEV as _jsxDEV } from "react/jsx-dev-runtime";

      export default function App() {
        return /*#__PURE__*/_jsxDEV(View, {}, void 0, false, { fileName: _jsxFileName, lineNumber: 4, columnNumber: 12 }, this);
      }"
    `);
  });
  it(`compiles for Webpack in prod`, () => {
    const options = {
      ...DEF_OPTIONS,
      presets: [[preset, { jsxRuntime: 'automatic' }]],
      caller: getCaller({
        name: 'babel-loader',
        isDev: false,
      }),
    };

    const code = babel.transform(sourceCode, options)!.code;

    expect(code).toMatch(/"react\/jsx-runtime"/);
    expect(code).toMatch(/_jsx\(View/);
    expect(code).toMatchInlineSnapshot(`
      "import View from "react-native-web/dist/exports/View";import { jsx as _jsx } from "react/jsx-runtime";

      export default function App() {
        return /*#__PURE__*/_jsx(View, {});
      }"
    `);
  });

  ['ios', 'web'].forEach((platform) => {
    it(`compiles for Metro ${platform} in dev`, () => {
      const options = {
        ...DEF_OPTIONS,
        presets: [[preset, { jsxRuntime: 'automatic' }]],
        caller: getCaller({
          name: 'metro',
          platform,
          isDev: true,
        }),
      };

      const code = babel.transform(sourceCode, options)!.code;

      expect(code).toMatch(/"react\/jsx-dev-runtime"/);
      // Format is a little different for Metro
      expect(code).not.toMatch(/_jsxRuntime\.jsx/);
      expect(code).toMatchSnapshot();
    });
    it(`compiles for Metro ${platform} in prod`, () => {
      const options = {
        ...DEF_OPTIONS,
        presets: [[preset, { jsxRuntime: 'automatic' }]],
        caller: getCaller({
          name: 'metro',
          platform,
          isDev: false,
        }),
      };

      const code = babel.transform(sourceCode, options)!.code;

      expect(code).toMatch(/"react\/jsx-runtime"/);
      // Format is a little different for Metro
      expect(code).toMatch(/_jsxRuntime\.jsx/);
      expect(code).toMatchSnapshot();
    });
  });
});
