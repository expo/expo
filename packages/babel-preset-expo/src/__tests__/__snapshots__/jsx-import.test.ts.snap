// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auto runtime (default) compiles for Metro ios in dev 1`] = `
"Object.defineProperty(exports, "__esModule", { value: true });exports.default = App;
var _reactNative = require("react-native");var _jsxDevRuntime = require("react/jsx-dev-runtime");var _jsxFileName = "/unknown";
function App() {
  return /*#__PURE__*/(0, _jsxDevRuntime.jsxDEV)(_reactNative.View, {}, void 0, false, { fileName: _jsxFileName, lineNumber: 4, columnNumber: 12 }, this);
}"
`;

exports[`auto runtime (default) compiles for Metro ios in prod 1`] = `
"Object.defineProperty(exports, "__esModule", { value: true });exports.default = App;
var _reactNative = require("react-native");var _jsxRuntime = require("react/jsx-runtime");
function App() {
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_reactNative.View, {});
}"
`;

exports[`auto runtime (default) compiles for Metro web in dev 1`] = `
"var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");Object.defineProperty(exports, "__esModule", { value: true });exports.default = App;var _View = _interopRequireDefault(require("react-native-web/dist/exports/View"));var _jsxDevRuntime = require("react/jsx-dev-runtime");var _jsxFileName = "/unknown";

function App() {
  return /*#__PURE__*/(0, _jsxDevRuntime.jsxDEV)(_View.default, {}, void 0, false, { fileName: _jsxFileName, lineNumber: 4, columnNumber: 12 }, this);
}"
`;

exports[`auto runtime (default) compiles for Metro web in prod 1`] = `
"var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");Object.defineProperty(exports, "__esModule", { value: true });exports.default = App;var _View = _interopRequireDefault(require("react-native-web/dist/exports/View"));var _jsxRuntime = require("react/jsx-runtime");

function App() {
  return /*#__PURE__*/(0, _jsxRuntime.jsx)(_View.default, {});
}"
`;

exports[`classic runtime compiles for Metro ios in dev 1`] = `
"Object.defineProperty(exports, "__esModule", { value: true });exports.default = App;
var _reactNative = require("react-native");var _jsxFileName = "/unknown";
function App() {
  return /*#__PURE__*/React.createElement(_reactNative.View, { __self: this, __source: { fileName: _jsxFileName, lineNumber: 4, columnNumber: 12 } });
}"
`;

exports[`classic runtime compiles for Metro ios in prod 1`] = `
"Object.defineProperty(exports, "__esModule", { value: true });exports.default = App;
var _reactNative = require("react-native");
function App() {
  return /*#__PURE__*/React.createElement(_reactNative.View, null);
}"
`;

exports[`classic runtime compiles for Metro web in dev 1`] = `
"var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");Object.defineProperty(exports, "__esModule", { value: true });exports.default = App;var _View = _interopRequireDefault(require("react-native-web/dist/exports/View"));var _jsxFileName = "/unknown";

function App() {
  return /*#__PURE__*/React.createElement(_View.default, { __self: this, __source: { fileName: _jsxFileName, lineNumber: 4, columnNumber: 12 } });
}"
`;

exports[`classic runtime compiles for Metro web in prod 1`] = `
"var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");Object.defineProperty(exports, "__esModule", { value: true });exports.default = App;var _View = _interopRequireDefault(require("react-native-web/dist/exports/View"));

function App() {
  return /*#__PURE__*/React.createElement(_View.default, null);
}"
`;
